### ISSUE TITLE
Revisiting the concept of "projections"

### ISSUE THREAD

**Issue by danielballan** [bluesky/tiled#399]
Before we starting thinking seriously about services, we added a concept of "projections" to Databroker. A "projection" is a JSON document describing how to remap the names and/or rearrange the contents of BlueskyRun or xarray into a different structure. This could be used, for example, to associate pieces of data and metadata in a BlueskyRun with items in a NeXus standard layout. Databroker provided this association (the "projection") as a kind of descriptive language. It was up to some "client-side" code to do something with it, via a "projector".

What happens when we revisit this idea from the perspective of a service? Now we care about:

* Clients that aren't written in Python (and cannot run Python "projectors")
* Generic HTTP clients that we didn't write, such as `curl` or web browsers

This suggests that the _server_ should do the rearrangement just ship the result. This also implies that it is not necessary to define a standard language for "projections". The server will run imperative code to perform the rearrangement. A server administrator _may_ define an internal system involving a descriptive language like projections to manage this, but that is an internal detail---not something Tiled has to expose or interact with directly.

This could be done today by registering additional formats. For example, I could define an `FXITomography` spec and assign that spec to certain BlueskyRuns from the FXI beamline. I could then register a custom exporter on the `FXITomography` spec that writes out a DataExchange structure in encoded HDF5. I could register a second exporter than does the same encoded in nested JSON.

Question for consideration: Should we pull apart the idea of serialization and restructuring, such that the choice of "DataExchange layout" vs "HDF5 or JSON" are separate parameters? This would involve adding a new query parameter (`transform`?) and a separate registry, very similar to our serialization registry but for _restructuring_ rather than serializiation. There is a risk that such a feature could be abused into become a remote data reduction/processing service. Transforming FXI tomography into DataExchange would in fact involve some light computation (interpolating angles). But as long as these transforms are parameter-free, I think this could potentially be used effectively. The amount of computation involved isn't really any more than the computation involved in serializing numpy arrays into JSON.

Worth some thought.

**Comment by dylanmcreynolds** [comment]
Tagging @ronpandolfi here, too. He is one of the known users of databroker v1 projections/projectors, so likely has some useful feedback about this.

**Comment by pshafer-als** [comment]
I don't know if it's already been ruled out here, but a while back @dylanmcreynolds and I were discussing a similar issue for another project: Using the HTTP header field 'Accept: _media-type_' instead of a query parameter to a request a particular representation that the server already knows about.

**Comment by danielballan** [comment]
Yeah, the `Accept` header is a standard and in a sense "preferred" way to specify the format, but it is convenient to be able to stick the `format` a URL, which can be more easily shared. I'm impressed you dug that out of where it's hidden in the docs...we could stand to improve documentation of the HTTP API overall.

**Comment by danielballan** [comment]
A good term for this parameter may be `layout`. We often use that in presentations when distinguishing serialization (on disk or in flight) from structure.
